src/kernel/kmain.zig

```diff
@@ -17,17 +17,89 @@ const MEMINFO = 1 << 1;
 const MAGIC = 0x1BADB002;
 const FLAGS = ALIGN | MEMINFO;
 
-export var multiboot align(4) linksection(".multiboot") = MultiBoot {
+export var multiboot align(4) linksection(".rodata.boot") = MultiBoot{
     .magic = MAGIC,
     .flags = FLAGS,
     .checksum = -(MAGIC + FLAGS),
 };
 
-export var stack_bytes: [16 * 1024]u8 align(16) linksection(".bss") = undefined;
-const stack_bytes_slice = stack_bytes[0..];
+const KERNEL_ADDR_OFFSET = 0xC0000000;
+const KERNEL_PAGE_NUMBER = KERNEL_ADDR_OFFSET >> 22;
+// The number of pages occupied by the kernel, will need to be increased as we add a heap etc.
+const KERNEL_NUM_PAGES = 1;
 
-export nakedcc fn _start() noreturn {
-    @newStackCall(stack_bytes_slice, kmain);
+// The initial page directory used for booting into the higher half. Should be overwritten later
+export var boot_page_directory: [1024]u32 align(4096) linksection(".rodata.boot") = init: {
+    // Temp value
+    var dir: [1024]u32 = undefined;
+
+    // Page for 0 -> 4 MiB. Gets unmapped later
+    dir[0] = 0x00000083;
+
+    var i = 0;
+    var idx = 1;
+
+    // Fill preceding pages with zeroes. May not be unecessary but incurs no runtime cost
+    while (i < KERNEL_PAGE_NUMBER - 1) {
+        dir[idx] = 0;
+        i += 1;
+        idx += 1;
+    }
+
+    // Map the kernel's higher half pages increasing by 4 MiB every time
+    i = 0;
+    while (i < KERNEL_NUM_PAGES) {
+        dir[idx] = 0x00000083 | (i << 22);
+        i += 1;
+        idx += 1;
+    }
+    // Increase max number of branches done by comptime evaluator
+    @setEvalBranchQuota(1024);
+    // Fill suceeding pages with zeroes. May not be unecessary but incurs no runtime cost
+    i = 0;
+    while (i < 1024 - KERNEL_PAGE_NUMBER - KERNEL_NUM_PAGES) {
+        dir[idx] = 0;
+        i += 1;
+        idx += 1;
+    }
+    break :init dir;
+};
+
+export var kernel_stack: [16 * 1024]u8 align(16) linksection(".bss.stack") = undefined;
+
+export nakedcc fn _start() align(16) linksection(".text.boot") noreturn {
+    // Seth the page directory to the boot directory
+    asm volatile (
```

Comment:
I don't know whether it is worth having these in there own function in a arch/x86.zig file as these are x86 specific

Reply:
That's the end goal, though I was planning on making another ticket where we go through the source and make a nice separation based on architecture.

Reply:
Could do, but don't want to get out of hand with, "I'll do it later" ;)